lib/Dialect/FIRRTL/Transforms/DropName.cpp

```diff
@@ -0,0 +1,80 @@
+//===- DropName.cpp - Drop Names  -----------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the DropName pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "llvm/ADT/STLExtras.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace firrtl;
+
+static bool isDeadOp(Operation *op,
+                     llvm::SmallDenseSet<Operation *> &deadOperations) {
+  for (auto *user : op->getUsers()) {
+    if (deadOperations.count(user))
+      continue;
+
+    auto connect = dyn_cast<FConnectLike>(user);
+    // If the user is neither node nor connect, we consider the op to be alive.
+    if (!connect)
+      return false;
+    if (connect.src().getDefiningOp() != op)
+      continue;
+    auto *destOp = connect.dest().getDefiningOp();
+    if (!destOp)
+      return false;
+    // Check the connect is self-connection. If not, check whether `dest` is
+    // known to be dead.
+    if (destOp != op && !deadOperations.count(destOp))
+      return false;
+  }
+
+  return true;
+}
+
+namespace {
+struct DropNamePass : public DropNameBase<DropNamePass> {
+  DropNamePass(bool dropOnlyDeadNameFlag) {
+    dropOnlyDeadNames = dropOnlyDeadNameFlag;
+  }
+  void runOnOperation() override {
+    // If `dropOnlyDeadNames` is false, we drop all names regardless of their
+    // liveness.
+    if (!dropOnlyDeadNames) {
+      getOperation().walk([](FNamableOp op) { op.dropName(); });
+      return;
+    }
+
+    llvm::SmallDenseSet<Operation *> deadOperations;
```

Comment:
This should happen via DCE and not be replicated here.

At minimum, LowerToHW could choose not to lower dead wires with droppable names.  (I think it already is testing for liveness)

Reply:
Yes, I agree that it is not ideal but I added this to avoid drastic changes on output verilog (https://github.com/llvm/circt/pull/3254#discussion_r888195934).  

To reduce LINT warnings, it is not sufficient not to lower dead operations at LowerToHW because of the operations that are transitively dead, e.g:
```scala
node a = input_port
node b = a
node c = b
node d = c
; d is head
```